### PR TITLE
fix(46149): Add specific top level await error messages

### LINKED ISSUE #46149: Provide a specialized error on top-level await and --module node12
Under `node12`, a top-level await results in

```
Top-level 'await' expressions are only allowed when the 'module' option is set to 'es2022', 'esnext', 'system', or 'nodenext', and the 'target' option is set to 'es2017' or higher.
```

This error message has too much text. We should consider specializing these cases for

* ES modules < es2022
* Node modules < nodenext

```
The module setting '{0}' does not support top-level 'await' expressions. Consider switching to '{1}'.
```

### PR DESCRIPTION
**Pull request by sidharthv96**
Fixes #46149

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -32183,10 +32183,23 @@ namespace ts {
...
+                            if (isNotTopLevelAwaitSupportedModuleKind(moduleKind, getSourceFileOfNode(node).impliedNodeFormat)) {
+                                if(moduleKind !== ModuleKind.NodeNext) {
+                                    const suggestedKind = moduleKind === ModuleKind.Node12 ? ModuleKind.NodeNext : ModuleKind.ES2022;
+                                    const diagnostic = createFileDiagnostic(sourceFile, span.start, span.length,
+                                        Diagnostics.The_0_setting_1_does_not_support_top_level_await_expressions_Consider_switching_to_2, "module", ModuleKind[moduleKind], ModuleKind[suggestedKind]);
+                                    diagnostics.add(diagnostic);
+                                }
+                                else {
+                                    const diagnostic = createFileDiagnostic(sourceFile, span.start, span.length,
+                                        Diagnostics._0_is_not_allowed_in_CommonJS_modules_Please_convert_to_ES_module, "Top-level 'await' expression");
```
Comment (weswigham) [inline src/compiler/checker.ts:32196]: This is probably bad for localization - I'd make this not take an argument and just make the whole message a specific diagnostic.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -42867,9 +42880,23 @@ namespace ts {
...
+                            if (isNotTopLevelAwaitSupportedModuleKind(moduleKind, getSourceFileOfNode(forInOrOfStatement).impliedNodeFormat)) {
+                                if(moduleKind !== ModuleKind.NodeNext) {
+                                    const suggestedKind = moduleKind === ModuleKind.Node12 ? ModuleKind.NodeNext : ModuleKind.ES2022;
+                                    const diagnostic = createDiagnosticForNode(forInOrOfStatement.awaitModifier,
+                                        Diagnostics.The_0_setting_1_does_not_support_top_level_for_await_loops_Consider_switching_to_2, "module", ModuleKind[moduleKind], ModuleKind[suggestedKind]);
+                                    diagnostics.add(diagnostic);
+                                }
+                                else {
+                                    const diagnostic = createDiagnosticForNode(forInOrOfStatement.awaitModifier,
+                                        Diagnostics._0_is_not_allowed_in_CommonJS_modules_Please_convert_to_ES_module, "Top-level 'for await' loop");
```
Comment (weswigham) [inline src/compiler/checker.ts:42892]: Likewise, this should probably just be a separate diagnostic for localization reasons.

File: src/compiler/types.ts [hunk 3]
```diff
@@ -6247,7 +6247,7 @@ namespace ts {
         Deferred = 7
     }
 
-    export const enum ScriptTarget {
+    export enum ScriptTarget {
```
Comment (weswigham) [inline src/compiler/types.ts:6250]: Leave this as `const` and just cast the ts namespac to `any` when doing the reverse mapping to silence the error about using a const enum (so `(ts as any).ScriptTarget[value]`). Don't worry, we preserve and un-const the const enums in our output, we do this elsewhere. 😊

**Review by weswigham — CHANGES_REQUESTED** [review]

**Comment by sandersn** [comment]
Do you want to keep working on this? Otherwise I'd like to close it to reduce the number of open PRs.

**Comment by sidharthv96** [comment]
Sorry, got really busy @work. Will push the changes over the weekend.

File: src/compiler/types.ts (hunk 3 above)
Comment (sidharthv96) [reply src/compiler/types.ts:6250]: Done, thanks!

File: src/compiler/checker.ts (hunk 2 above)
Comment (sidharthv96) [reply src/compiler/checker.ts]: Done.

File: src/compiler/checker.ts (hunk 1 above)
Comment (sidharthv96) [reply src/compiler/checker.ts]: Thanks, done.

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -815,6 +815,10 @@ namespace ts {
         return isExternalModule(node) || compilerOptions.isolatedModules || (isCommonJSContainingModuleKind(getEmitModuleKind(compilerOptions)) && !!node.commonJsModuleIndicator);
     }
 
+    export function isNotTopLevelAwaitSupportedModuleKind(kind: ModuleKind, impliedNodeFormat?: ModuleKind.ESNext | ModuleKind.CommonJS){
+        return kind !== ModuleKind.ES2022 && kind !== ModuleKind.ESNext && kind !== ModuleKind.System && !(kind === ModuleKind.NodeNext && impliedNodeFormat === ModuleKind.ESNext);
```
Comment (weswigham) [inline src/compiler/utilities.ts:819]: We bumped from `node12` to `node14`, specifically for TLA, so this needs to be updated to account for that.

**Review by weswigham — COMMENTED** [review]
You know what, between the major refactoring to the checker and us removing `node12` and replacing it with `node14`, I think this needs some changes at this point to make sense. I'm going to close this PR - if you or anyone else wants to take another crack at better TLA errors in this new context, feel free to open a new PR - that way it'll get back on the pile of things for us to look at faster. Thanks for the help 💖

**Pull request closed without merging**